File: MainWindowTabKaraokeMerge.vb
Imports Subtitle_draft_GMTPC.Services

Partial Class MainWindow

#Region "Karaoke Merge - Fields"

    Private _isKaraokeMergeUpdating As Boolean = False

#End Region

#Region "Karaoke Merge - Event Handlers"

    Private Sub TxtKaraokeMergeInput_TextChanged(sender As Object, e As TextChangedEventArgs)
        If _isKaraokeMergeUpdating Then Return
        Try
            _isKaraokeMergeUpdating = True
            Dim content = SubtitleParser.SanitizeContent(TxtKaraokeMergeInput.Text)
            If String.IsNullOrWhiteSpace(content) Then
                TxtKaraokeMergeCount.Text = ""
                TxtKaraokeMergeOutput.Text = ""
                Return
            End If

            Dim lines = content.Split({Environment.NewLine, vbCr, vbLf}, StringSplitOptions.RemoveEmptyEntries)
            Dim dialogueCount = lines.Count(Function(l) l.Trim().StartsWith("Dialogue:"))
            TxtKaraokeMergeCount.Text = String.Format("({0} dialogue lines)", dialogueCount)

            ' Xu ly merge karaoke
            TxtKaraokeMergeOutput.Text = KaraokeMergeService.ProcessKaraokeMerge(content)
        Catch ex As Exception
            TxtKaraokeMergeCount.Text = String.Format("(Error: {0})", ex.Message)
        Finally
            _isKaraokeMergeUpdating = False
        End Try
    End Sub

#End Region

#Region "Karaoke Merge - Toast"

    Private Async Sub ShowToastKaraokeMerge(message As String)
        ToastTextKaraokeMerge.Text = message
        ToastBorderKaraokeMerge.Visibility = Visibility.Visible
        Await Task.Delay(2000)
        ToastBorderKaraokeMerge.Visibility = Visibility.Collapsed
    End Sub

#End Region

#Region "Karaoke Merge - Copy Button"

    Private Sub BtnCopyKaraokeMerge_Click(sender As Object, e As RoutedEventArgs)
        If String.IsNullOrWhiteSpace(TxtKaraokeMergeOutput.Text) Then Return
        Try
            Clipboard.SetText(TxtKaraokeMergeOutput.Text)
            ShowToastKaraokeMerge("📋 Copied Merged Karaoke!")
        Catch ex As Exception
            MessageBox.Show("Error: " & ex.Message, "Error", MessageBoxButton.OK, MessageBoxImage.Error)
        End Try
    End Sub

#End Region

End Class

File: MainWindowTabKaraokeMerge.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Windows;
using System.Windows.Controls;
using System.Threading.Tasks;
using Subtitle_draft_GMTPC.Models;
using Subtitle_draft_GMTPC.Services;

namespace Subtitle_draft_GMTPC
{
    public partial class MainWindow : Window
    {

        #region Karaoke Merge - Fields

        private bool _isKaraokeMergeUpdating = false;

        #endregion

        #region Karaoke Merge - Event Handlers

        private void TxtKaraokeMergeInput_TextChanged(object sender, TextChangedEventArgs e)
        {
            if (_isKaraokeMergeUpdating) return;
            try
            {
                _isKaraokeMergeUpdating = true;
                var content = SubtitleParser.SanitizeContent(TxtKaraokeMergeInput.Text);
                if (string.IsNullOrWhiteSpace(content))
                {
                    TxtKaraokeMergeCount.Text = "";
                    TxtKaraokeMergeOutput.Text = "";
                    return;
                }

                var lines = content.Split(new[] { Environment.NewLine, "\r", "\n" }, StringSplitOptions.RemoveEmptyEntries);
                var dialogueCount = lines.Count(l => l.Trim().StartsWith("Dialogue:"));
                TxtKaraokeMergeCount.Text = string.Format("({0} dialogue lines)", dialogueCount);

                // Xu ly merge karaoke
                TxtKaraokeMergeOutput.Text = KaraokeMergeService.ProcessKaraokeMerge(content);
            }
            catch (Exception ex)
            {
                TxtKaraokeMergeCount.Text = string.Format("(Error: {0})", ex.Message);
            }
            finally
            {
                _isKaraokeMergeUpdating = false;
            }
        }

        #endregion

        #region Karaoke Merge - Toast

        private async void ShowToastKaraokeMerge(string message)
        {
            ToastTextKaraokeMerge.Text = message;
            ToastBorderKaraokeMerge.Visibility = Visibility.Visible;
            await Task.Delay(2000);
            ToastBorderKaraokeMerge.Visibility = Visibility.Collapsed;
        }

        #endregion

        #region Karaoke Merge - Copy Button

        private void BtnCopyKaraokeMerge_Click(object sender, RoutedEventArgs e)
        {
            if (string.IsNullOrWhiteSpace(TxtKaraokeMergeOutput.Text)) return;
            try
            {
                Clipboard.SetText(TxtKaraokeMergeOutput.Text);
                ShowToastKaraokeMerge("📋 Copied Merged Karaoke!");
            }
            catch (Exception ex)
            {
                System.Windows.MessageBox.Show("Error: " + ex.Message, "Error", System.Windows.MessageBoxButton.OK, System.Windows.MessageBoxImage.Error);
            }
        }

        #endregion

    }
}
